File: src/lib/sound.js
// WebAudio sound kit. The buzzers, dings and flips are synthesised, so there
// are no files for those; the audience reactions are real recordings from
// public/audio (see tools/crowd/). voice.js plays through the same graph.

let ctx = null;
let master = null;
let room = null;

let muted = (() => {
  try {
    return localStorage.getItem("cs-muted") === "1";
  } catch {
    return false;
  }
})();

export const isMuted = () => muted;
export function setMuted(m) {
  muted = m;
  if (master) master.gain.value = m ? 0 : 1;
  try {
    localStorage.setItem("cs-muted", m ? "1" : "0");
  } catch {
    /* storage unavailable */
  }
}

/** The shared context, created on first use (must follow a user gesture). */
export function audio() {
  if (!ctx) {
    ctx = new (window.AudioContext || window.webkitAudioContext)();
    master = ctx.createGain();
    master.gain.value = muted ? 0 : 1;
    master.connect(ctx.destination);
    // A short synthetic room reverb: makes voices sound like a studio, not a phone.
    room = ctx.createConvolver();
    const len = ctx.sampleRate * 1.6;
    const ir = ctx.createBuffer(2, len, ctx.sampleRate);
    for (let ch = 0; ch < 2; ch++) {
      const d = ir.getChannelData(ch);
      for (let i = 0; i < len; i++) d[i] = (Math.random() * 2 - 1) * Math.pow(1 - i / len, 3);
    }
    room.buffer = ir;
    const wet = ctx.createGain();
    wet.gain.value = 0.22;
    room.connect(wet).connect(master);
  }
  if (ctx.state === "suspended") ctx.resume();
  return { ctx, master, room };
}

function tone(freq, start, dur, type = "sine", gain = 0.18) {
  if (muted) return;
  const { ctx, master } = audio();
  const t = ctx.currentTime + start;
  const osc = ctx.createOscillator();
  const g = ctx.createGain();
  osc.type = type;
  osc.frequency.setValueAtTime(freq, t);
  g.gain.setValueAtTime(0, t);
  g.gain.linearRampToValueAtTime(gain, t + 0.01);
  g.gain.exponentialRampToValueAtTime(0.0001, t + dur);
  osc.connect(g).connect(master);
  osc.start(t);
  osc.stop(t + dur + 0.05);
}

let noiseBuf = null;
function noise() {
  const { ctx } = audio();
  if (!noiseBuf) {
    noiseBuf = ctx.createBuffer(1, ctx.sampleRate * 2, ctx.sampleRate);
    const d = noiseBuf.getChannelData(0);
    for (let i = 0; i < d.length; i++) d[i] = Math.random() * 2 - 1;
  }
  const src = ctx.createBufferSource();
  src.buffer = noiseBuf;
  src.loop = true;
  return src;
}

/*
  Real studio audiences, cut from freely-licensed recordings (see
  tools/crowd/ATTRIBUTION.md). Synthesised noise can do the texture of a room
  but never the sound of actual people, so reactions play the recordings and
  fall back to the synth below only if a clip can't be loaded.
*/
const SAMPLES = {
  cheer: "/audio/cheer.m4a",
  applause: "/audio/applause.m4a",
  applauseBig: "/audio/applause-big.m4a",
  groan: "/audio/groan.m4a",
};
const loaded = new Map();

function sample(name) {
  if (!loaded.has(name)) {
    const p = fetch(SAMPLES[name])
      .then((r) => {
        if (!r.ok) throw new Error(`audio ${r.status}`);
        return r.arrayBuffer();
      })
      .then((ab) => audio().ctx.decodeAudioData(ab));
    p.catch(() => loaded.delete(name));
    loaded.set(name, p);
  }
  return loaded.get(name);
}

/** Fetch and decode the reactions up front, so the first one isn't late. */
export function preloadCrowd() {
  if (muted) return; // needs a gesture first; called again once unmuted
  for (const name of Object.keys(SAMPLES)) sample(name).catch(() => {});
}

/** Plays a recording at `level`, trimmed to `dur` with a short fade. */
function playSample(name, dur, level, { wet = true } = {}) {
  if (muted) return false;
  const p = loaded.get(name);
  if (!p) {
    sample(name).catch(() => {});
    return false; // not decoded yet — this one uses the synth
  }
  p.then((buf) => {
    if (muted) return;
    const { ctx, master, room } = audio();
    const src = ctx.createBufferSource();
    src.buffer = buf;
    const g = ctx.createGain();
    const t = ctx.currentTime;
    const length = Math.min(dur ?? buf.duration, buf.duration);
    g.gain.setValueAtTime(0, t);
    g.gain.linearRampToValueAtTime(level, t + 0.04);
    g.gain.setValueAtTime(level, t + Math.max(0.05, length - 0.4));
    g.gain.linearRampToValueAtTime(0.0001, t + length);
    src.connect(g).connect(master);
    if (wet) g.connect(room);
    src.start(t);
    src.stop(t + length + 0.05);
  }).catch(() => {});
  return true;
}

/** Hundreds of tiny filtered noise bursts read convincingly as a clapping crowd. */
function synthApplause(dur = 2.6, level = 0.5) {
  if (muted) return;
  const { ctx, master, room } = audio();
  const out = ctx.createGain();
  out.gain.value = level;
  out.connect(master);
  out.connect(room);
  const now = ctx.currentTime;
  const claps = Math.floor(dur * 140);
  for (let i = 0; i < claps; i++) {
    const t = now + Math.random() * dur;
    // Fade the density out towards the end.
    const fade = 1 - Math.max(0, (t - now) / dur - 0.55) / 0.45;
    const src = noise();
    const bp = ctx.createBiquadFilter();
    bp.type = "bandpass";
    bp.frequency.value = 900 + Math.random() * 1800;
    bp.Q.value = 1.2;
    const g = ctx.createGain();
    g.gain.setValueAtTime(0, t);
    g.gain.linearRampToValueAtTime(0.25 * fade, t + 0.002);
    g.gain.exponentialRampToValueAtTime(0.0001, t + 0.05);
    src.connect(bp).connect(g).connect(out);
    src.start(t, Math.random());
    src.stop(t + 0.06);
  }
}

function applause(dur = 2.6, level = 0.5) {
  if (muted) return;
  // A four-second-plus cue is a win, so bring the whole house in.
  const name = dur >= 4 ? "applauseBig" : "applause";
  if (!playSample(name, dur, Math.min(1, level * 1.5))) synthApplause(dur, level);
}

/** Crowd bed: a real audience under the voices, or filtered noise if it's late. */
export function crowdBed(dur = 1.4, level = 0.1, shape = "cheer") {
  if (muted) return;
  if (playSample(shape === "groan" ? "groan" : "cheer", dur, Math.min(1, level * 3.2))) return;
  const { ctx, master } = audio();
  const src = noise();
  const bp = ctx.createBiquadFilter();
  bp.type = "bandpass";
  bp.frequency.value = shape === "groan" ? 420 : 1100;
  bp.Q.value = 0.7;
  const g = ctx.createGain();
  const t = ctx.currentTime;
  g.gain.setValueAtTime(0, t);
  g.gain.linearRampToValueAtTime(level, t + 0.12);
  g.gain.setValueAtTime(level, t + dur * 0.6);
  g.gain.exponentialRampToValueAtTime(0.0001, t + dur);
  if (shape === "groan") {
    bp.frequency.setValueAtTime(520, t);
    bp.frequency.exponentialRampToValueAtTime(260, t + dur);
  }
  src.connect(bp).connect(g).connect(master);
  src.start(t);
  src.stop(t + dur + 0.1);
}

export const sfx = {
  ding() {
    tone(1318, 0, 0.5, "triangle", 0.22);
    tone(1760, 0.09, 0.7, "triangle", 0.18);
  },
  buzz() {
    tone(110, 0, 0.7, "sawtooth", 0.16);
    tone(116, 0, 0.7, "square", 0.08);
  },
  flip() {
    tone(520, 0, 0.08, "square", 0.05);
    tone(880, 0.05, 0.14, "triangle", 0.1);
  },
  dupe() {
    tone(330, 0, 0.12, "square", 0.1);
    tone(330, 0.16, 0.12, "square", 0.1);
  },
  tick() {
    tone(1200, 0, 0.04, "square", 0.05);
  },
  buzzer() {
    // Face-off buzz-in.
    tone(740, 0, 0.25, "square", 0.12);
    tone(988, 0, 0.25, "square", 0.06);
  },
  timeUp() {
    tone(220, 0, 1.2, "sawtooth", 0.14);
    tone(233, 0, 1.2, "square", 0.08);
  },
  win() {
    [523, 659, 784, 1047].forEach((f, i) => tone(f, i * 0.12, 0.45, "triangle", 0.18));
  },
  applause,
  groan() {
    if (!playSample("groan", 1.8, 0.55)) crowdBed(1.3, 0.16, "groan");
  },
};

File: ios/CrowdSays/Lib/Sound.swift
import AVFoundation
import Foundation

/*
  The sound kit. Every tone, clap and crowd murmur is synthesised into a PCM
  buffer at runtime, so the app ships no audio files — the same approach
  src/lib/sound.js takes with WebAudio. Voice.swift plays through this graph
  too, which is what puts the host and the studio audience in the same room.
*/

enum Wave { case sine, triangle, square, sawtooth }

/// One component of a sound effect.
struct ToneSpec {
    var freq: Double
    var start: Double = 0
    var dur: Double
    var wave: Wave = .sine
    var gain: Double = 0.18
}

private let sr = 44_100.0

// ── Buffer rendering ──

private func makeBuffer(_ seconds: Double) -> (AVAudioPCMBuffer, UnsafeMutablePointer<Float>)? {
    let frames = AVAudioFrameCount(max(1, seconds * sr))
    guard let fmt = AVAudioFormat(standardFormatWithSampleRate: sr, channels: 1),
          let buf = AVAudioPCMBuffer(pcmFormat: fmt, frameCapacity: frames),
          let data = buf.floatChannelData?[0]
    else { return nil }
    buf.frameLength = frames
    data.update(repeating: 0, count: Int(frames))
    return (buf, data)
}

private func sample(_ wave: Wave, _ phase: Double) -> Double {
    let t = phase - phase.rounded(.down) // 0..<1
    switch wave {
    case .sine: return sin(2 * .pi * t)
    case .square: return t < 0.5 ? 1 : -1
    case .sawtooth: return 2 * t - 1
    case .triangle: return 4 * abs(t - 0.5) - 1
    }
}

enum Synth {
    /// Renders tones that share one buffer, each with the WebAudio envelope:
    /// a 10ms attack then an exponential decay to silence.
    static func tones(_ specs: [ToneSpec]) -> AVAudioPCMBuffer? {
        guard let total = specs.map({ $0.start + $0.dur + 0.05 }).max(),
              let (buf, data) = makeBuffer(total)
        else { return nil }

        for spec in specs {
            let from = Int(spec.start * sr)
            let to = min(Int(buf.frameLength), Int((spec.start + spec.dur) * sr))
            guard to > from else { continue }
            let attack = 0.01 * sr
            for i in from..<to {
                let n = Double(i - from)
                let env: Double
                if n < attack {
                    env = spec.gain * (n / attack)
                } else {
                    // exponentialRampToValueAtTime(0.0001, start + dur)
                    let p = n / (spec.dur * sr)
                    env = spec.gain * pow(0.0001 / max(spec.gain, 0.0001), p)
                }
                data[i] += Float(env * sample(spec.wave, n / sr * spec.freq))
            }
        }
        return buf
    }

    /// Hundreds of tiny filtered noise bursts read convincingly as a clapping
    /// crowd. The density fades towards the end, as it does on the web.
    static func applause(dur: Double = 2.6, level: Double = 0.5) -> AVAudioPCMBuffer? {
        guard let (buf, data) = makeBuffer(dur + 0.2) else { return nil }
        let claps = Int(dur * 140)
        for _ in 0..<claps {
            let at = Double.random(in: 0..<dur)
            let fade = min(1, max(0, 1 - max(0, at / dur - 0.55) / 0.45))
            let start = Int(at * sr)
            let len = Int(0.05 * sr)
            // A single-pole bandpass on white noise, cheap and close enough.
            let f = Double.random(in: 900...2700) / sr
            var prev = 0.0
            var band = 0.0
            for i in 0..<len {
                let idx = start + i
                if idx >= Int(buf.frameLength) { break }
                let white = Double.random(in: -1...1)
                band += f * (white - prev - band)
                prev += f * band
                let env = 0.25 * fade * pow(0.0001 / 0.25, Double(i) / Double(len))
                data[idx] += Float(prev * env * level * 2)
            }
        }
        return buf
    }

    /// Linear-resamples a buffer to detune it. AVAudioPlayerNode has no rate
    /// control, so a layered crowd is built by resampling each voice instead.
    static func resample(_ buf: AVAudioPCMBuffer, rate: Double) -> AVAudioPCMBuffer {
        guard abs(rate - 1) > 0.001, rate > 0,
              let src = buf.floatChannelData
        else { return buf }
        let channels = Int(buf.format.channelCount)
        let inFrames = Int(buf.frameLength)
        let outFrames = Int(Double(inFrames) / rate)
        guard outFrames > 1,
              let out = AVAudioPCMBuffer(pcmFormat: buf.format, frameCapacity: AVAudioFrameCount(outFrames)),
              let dst = out.floatChannelData
        else { return buf }
        out.frameLength = AVAudioFrameCount(outFrames)
        for ch in 0..<channels {
            for i in 0..<outFrames {
                let pos = Double(i) * rate
                let i0 = min(inFrames - 1, Int(pos))
                let i1 = min(inFrames - 1, i0 + 1)
                let frac = Float(pos - Double(i0))
                dst[ch][i] = src[ch][i0] * (1 - frac) + src[ch][i1] * frac
            }
        }
        return out
    }

    /// Crowd bed: bandpassed noise with a wobbling envelope, under the voices.
    static func crowdBed(dur: Double = 1.4, level: Double = 0.1, groan: Bool = false) -> AVAudioPCMBuffer? {
        guard let (buf, data) = makeBuffer(dur + 0.1) else { return nil }
        let frames = Int(buf.frameLength)
        var prev = 0.0
        var band = 0.0
        for i in 0..<frames {
            let t = Double(i) / sr
            // A groan slides the voice of the room down as it fades.
            let center = groan ? 520 * pow(260.0 / 520.0, min(1, t / dur)) : 1100.0
            let f = center / sr
            let white = Double.random(in: -1...1)
            band += f * (white - prev - band)
            prev += f * band

            let env: Double
            if t < 0.12 {
                env = level * (t / 0.12)
            } else if t < dur * 0.6 {
                env = level
            } else {
                let p = min(1, (t - dur * 0.6) / max(0.0001, dur * 0.4))
                env = level * pow(0.0001 / max(level, 0.0001), p)
            }
            data[i] = Float(prev * env * 3)
        }
        return buf
    }
}

// ── The engine ──

/// One layer of a scheduled playback.
struct Layer {
    var buffer: AVAudioPCMBuffer
    var delay: Double = 0
    var rate: Double = 1
    var gain: Double = 1
    var pan: Double = 0
    /// Whether this layer also feeds the room reverb.
    var wet: Bool = true
}

final class AudioEngine {
    static let shared = AudioEngine()

    private let engine = AVAudioEngine()
    private let dry = AVAudioMixerNode()
    private let wet = AVAudioMixerNode()
    private let reverb = AVAudioUnitReverb()
    private let lock = NSLock()
    private var started = false

    var isMuted: Bool = Store.isMuted {
        didSet {
            Store.isMuted = isMuted
            dry.outputVolume = isMuted ? 0 : 1
            wet.outputVolume = isMuted ? 0 : 0.22
        }
    }

    private init() {}

    /// Configures the audio session and graph on first use. Playback mode so
    /// the show keeps talking when the phone is on silent.
    func start() {
        lock.lock()
        defer { lock.unlock() }
        guard !started else { return }
        started = true

        #if os(iOS)
        // Playback category, so the show keeps talking when the ringer is off.
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.playback, mode: .default, options: [])
        try? session.setActive(true)
        #endif

        engine.attach(dry)
        engine.attach(wet)
        engine.attach(reverb)
        reverb.loadFactoryPreset(.mediumRoom)
        reverb.wetDryMix = 100

        let fmt = AVAudioFormat(standardFormatWithSampleRate: sr, channels: 2)
        engine.connect(dry, to: engine.mainMixerNode, format: fmt)
        engine.connect(wet, to: reverb, format: fmt)
        engine.connect(reverb, to: engine.mainMixerNode, format: fmt)
        wet.outputVolume = isMuted ? 0 : 0.22 // matches the web's wet gain
        dry.outputVolume = isMuted ? 0 : 1

        engine.prepare()
        try? engine.start()
    }

    /// Schedules every layer and returns how long the whole thing runs.
    @discardableResult
    func play(_ layers: [Layer]) -> Double {
        guard !isMuted, !layers.isEmpty else { return 0 }
        start()
        guard engine.isRunning else { return 0 }

        var end = 0.0
        for layer in layers {
            let buffer = Synth.resample(layer.buffer, rate: layer.rate)
            let node = AVAudioPlayerNode()
            engine.attach(node)
            let fmt = buffer.format
            // One output bus feeding both mixers: a second connect() call would
            // replace this connection rather than add to it.
            var points = [AVAudioConnectionPoint(node: dry, bus: dry.nextAvailableInputBus)]
            if layer.wet {
                points.append(AVAudioConnectionPoint(node: wet, bus: wet.nextAvailableInputBus))
            }
            engine.connect(node, to: points, fromBus: 0, format: fmt)
            node.volume = Float(layer.gain)
            node.pan = Float(layer.pan)

            let duration = Double(buffer.frameLength) / fmt.sampleRate
            end = max(end, layer.delay + duration)

            node.scheduleBuffer(buffer, at: nil, options: []) { [weak self] in
                // Detach on the main actor once the buffer has drained.
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                    guard let self else { return }
                    node.stop()
                    self.engine.detach(node)
                }
            }
            if layer.delay > 0 {
                DispatchQueue.main.asyncAfter(deadline: .now() + layer.delay) { node.play() }
            } else {
                node.play()
            }
        }
        return end
    }

    /// Plays and waits for the sound to finish, so the show can sequence on it.
    func playAndWait(_ layers: [Layer]) async {
        let end = play(layers)
        guard end > 0 else { return }
        try? await Task.sleep(nanoseconds: UInt64((end + 0.06) * 1_000_000_000))
    }

    func one(_ buffer: AVAudioPCMBuffer?, gain: Double = 1, wet: Bool = false) {
        guard let buffer else { return }
        play([Layer(buffer: buffer, gain: gain, wet: wet)])
    }
}

/*
  Real studio audiences, cut from freely-licensed recordings (see
  tools/crowd/ATTRIBUTION.md) and bundled with the app. Synthesised noise can
  do the texture of a room but never the sound of actual people, so reactions
  play the recordings and fall back to the synth only if one is missing.
*/
enum CrowdSample: String, CaseIterable {
    case cheer, applause, groan
    case applauseBig = "applause-big"

    /// Decoded once and kept: four short clips is nothing to hold in memory.
    private static var cache: [CrowdSample: AVAudioPCMBuffer] = [:]
    private static let lock = NSLock()

    var buffer: AVAudioPCMBuffer? {
        CrowdSample.lock.lock()
        defer { CrowdSample.lock.unlock() }
        if let hit = CrowdSample.cache[self] { return hit }
        guard let url = Bundle.main.url(forResource: rawValue, withExtension: "m4a"),
              let file = try? AVAudioFile(forReading: url),
              let buf = AVAudioPCMBuffer(pcmFormat: file.processingFormat,
                                         frameCapacity: AVAudioFrameCount(file.length)),
              (try? file.read(into: buf)) != nil
        else { return nil }
        CrowdSample.cache[self] = buf
        return buf
    }

    /// Decodes every clip up front, so the first reaction isn't late.
    static func preload() {
        Task.detached(priority: .utility) {
            for s in CrowdSample.allCases { _ = s.buffer }
        }
    }
}

// ── The kit, pre-rendered once ──

enum SFX {
    // These never change, so they're built on first use and reused.
    private static let dingBuf = Synth.tones([
        ToneSpec(freq: 1318, dur: 0.5, wave: .triangle, gain: 0.22),
        ToneSpec(freq: 1760, start: 0.09, dur: 0.7, wave: .triangle, gain: 0.18),
    ])
    private static let buzzBuf = Synth.tones([
        ToneSpec(freq: 110, dur: 0.7, wave: .sawtooth, gain: 0.16),
        ToneSpec(freq: 116, dur: 0.7, wave: .square, gain: 0.08),
    ])
    private static let flipBuf = Synth.tones([
        ToneSpec(freq: 520, dur: 0.08, wave: .square, gain: 0.05),
        ToneSpec(freq: 880, start: 0.05, dur: 0.14, wave: .triangle, gain: 0.1),
    ])
    private static let dupeBuf = Synth.tones([
        ToneSpec(freq: 330, dur: 0.12, wave: .square, gain: 0.1),
        ToneSpec(freq: 330, start: 0.16, dur: 0.12, wave: .square, gain: 0.1),
    ])
    private static let tickBuf = Synth.tones([ToneSpec(freq: 1200, dur: 0.04, wave: .square, gain: 0.05)])
    private static let buzzerBuf = Synth.tones([
        ToneSpec(freq: 740, dur: 0.25, wave: .square, gain: 0.12),
        ToneSpec(freq: 988, dur: 0.25, wave: .square, gain: 0.06),
    ])
    private static let timeUpBuf = Synth.tones([
        ToneSpec(freq: 220, dur: 1.2, wave: .sawtooth, gain: 0.14),
        ToneSpec(freq: 233, dur: 1.2, wave: .square, gain: 0.08),
    ])
    private static let winBuf = Synth.tones(
        [523.0, 659, 784, 1047].enumerated().map { i, f in
            ToneSpec(freq: f, start: Double(i) * 0.12, dur: 0.45, wave: .triangle, gain: 0.18)
        }
    )

    static func ding() { AudioEngine.shared.one(dingBuf, gain: 1) }
    static func buzz() { AudioEngine.shared.one(buzzBuf) }
    static func flip() { AudioEngine.shared.one(flipBuf) }
    static func dupe() { AudioEngine.shared.one(dupeBuf) }
    static func tick() { AudioEngine.shared.one(tickBuf) }
    static func buzzer() { AudioEngine.shared.one(buzzerBuf) }
    static func timeUp() { AudioEngine.shared.one(timeUpBuf) }
    static func win() { AudioEngine.shared.one(winBuf, wet: true) }

    /// Plays a recorded reaction, or returns false so the caller can synthesise.
    private static func reaction(_ s: CrowdSample, level: Double) -> Bool {
        guard let buf = s.buffer else { return false }
        AudioEngine.shared.play([Layer(buffer: buf, gain: min(1, level), wet: true)])
        return true
    }

    static func applause(_ dur: Double = 2.6, _ level: Double = 0.5) {
        guard !AudioEngine.shared.isMuted else { return }
        // A four-second-plus cue is a win, so bring the whole house in.
        if reaction(dur >= 4 ? .applauseBig : .applause, level: level * 1.5) { return }
        Task.detached(priority: .userInitiated) {
            let buf = Synth.applause(dur: dur, level: level)
            await MainActor.run { AudioEngine.shared.one(buf, wet: true) }
        }
    }

    static func crowdBed(_ dur: Double = 1.4, _ level: Double = 0.1, groan: Bool = false) {
        guard !AudioEngine.shared.isMuted else { return }
        if reaction(groan ? .groan : .cheer, level: level * 3.2) { return }
        Task.detached(priority: .userInitiated) {
            let buf = Synth.crowdBed(dur: dur, level: level, groan: groan)
            await MainActor.run { AudioEngine.shared.one(buf) }
        }
    }

    static func groan() {
        guard !AudioEngine.shared.isMuted else { return }
        if reaction(.groan, level: 0.55) { return }
        crowdBed(1.3, 0.16, groan: true)
    }
}
